pubsublite-spark-sql-streaming/src/main/java/com/google/cloud/pubsublite/spark/PslSourceOffset.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsublite.spark;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.Partition;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public final class PslSourceOffset extends org.apache.spark.sql.sources.v2.reader.streaming.Offset {
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  static {
+    objectMapper.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
+  }
+
+  private final ImmutableMap<Partition, Offset> partitionOffsetMap;
```

Comment:
Based on your other comments, you're planning on making these either Pub/Sub Lite offsets or spark open-closed offsets by context? This is error prone. Please ensure that anywhere these are set, they are Pub/Sub Lite offsets.

Reply:
Wrapping back around to this after reading spark docs: This is the primary type handled by spark for its open-closed ranges. Instead, always ensure this refers to the spark offset.

Reply:
Yes, PslSourceOffset and PslPartitionOffset are spark offsets implementations that wraps PSL partition and offsets. Once entering spark, those spark offsets are the ones passing around.